### ISSUE TITLE
Building on MacOS with custom librdkafka prefix

### ISSUE THREAD

**Issue by sylr** [confluentinc/confluent-kafka-go#420]
Description
===========

Hi,

I'm currently writing a go program that will rely on https://github.com/confluentinc/confluent-kafka-go.

For convenience `librdkafka` is shipped within my program repository in a git submodule located in `$(PROJECT_ROOT)/lib/librdkafka`. I build librdkafka with the following Makefile target:

```Makefile
librdkafka:
	git submodule update --init
	cd ./lib/librdkafka && ./configure --prefix="$(CURDIR)/build"
	make -C ./lib/librdkafka/
	make -C ./lib/librdkafka/ install
```

Once this is done I try to compile my project but it fails cause I can't get cgo to look for librdkafka in `$(CURDIR)/build/...`.
In fact I can but it tries really hard to look for it in `/usr/local/Cellar/librdkafka/1.3.0/lib` first and that makes `ld` fail.

```
GOOS=darwin GOARCH=amd64 go build -ldflags '-extldflags "-v -L/Users/s.rabot/git/go-lectra-maintenance/build/lib -lrdkafka"' -o dist/maintenance
# gitlab.lectra.com/cloud/maintenance
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin19.2.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/ld" -demangle -lto_library /Library/Developer/CommandLineTools/usr/lib/libLTO.dylib -dynamic -arch x86_64 -macosx_version_min 10.15.0 -syslibroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -o $WORK/b001/exe/a.out -L/usr/local/Cellar/librdkafka/1.3.0/lib -L/Users/s.rabot/git/go-lectra-maintenance/build/lib -headerpad 1144 -no_pie -pagezero_size 4000000 /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/go.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000000.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000001.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000002.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000003.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000004.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000005.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000006.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000007.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000008.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000009.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000010.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000011.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000012.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000013.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000014.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000015.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000016.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000017.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000018.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000019.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000020.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000021.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000022.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000023.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000024.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000025.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000026.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000027.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000028.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000029.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000030.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000031.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000032.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000033.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000034.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-619783463/000035.o -lrdkafka -lpthread -framework CoreFoundation -framework Security -lrdkafka -L/usr/local/lib -lSystem /Library/Developer/CommandLineTools/usr/lib/clang/11.0.0/lib/darwin/libclang_rt.osx.a
ld: warning: directory not found for option '-L/usr/local/Cellar/librdkafka/1.3.0/lib'
```

I used to have `librdkafka` installed with `brew` but I unstalled/removed/unlinked it.

I've spent a few hours trying to understand how cgo generates the list of directory used for linking but I did not succeed.

Hope you could help me with that.

Thanks.

Checklist
=========
Please provide the following information:

 - [ ] confluent-kafka-go and librdkafka version (`LibraryVersion()`): v1.3.0 / v1.3.0-13-g43b214-dirty
 - [ ] Operating system: MacOS 10.15.2
 - [ ] Provide client logs (with `"debug": ".."` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
confluent-kafka-go uses pkg-config to find the include and library paths of librdkafka, and since you're not passing/overriding PKGCONFIG_PATH to point to your build-directory's pc path it instead finds the librdkafka pkgconfig file installed by homebrew.

Try `PKGCONFIG_PATH=/full/path/to/build/lib/pkgconfig go build` (or wherever the rdkafka.pc file is installed).

In early Q2 we'll start bundling a prebuilt librdkafka with the Go client to remove the external dependency on librdkafka, you can try it out by referencing https://github.com/confluentinc/confluent-kafka-go-dev/

**Comment by sylr** [comment]
Hi @edenhill, thank you for your response.

I've tried `PKGCONFIG_PATH=/Users/s.rabot/git/go-lectra-maintenance/build/lib/pkgconfig` as you mentioned but the error still persists.

```
PKGCONFIG_PATH=/Users/s.rabot/git/go-lectra-maintenance/build/lib/pkgconfig GOOS=darwin GOARCH=amd64 go build -ldflags '-extldflags "-v -L/Users/s.rabot/git/go-lectra-maintenance/build/lib -lrdkafka"' -o dist/maintenance
# gitlab.lectra.com/cloud/maintenance
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin19.2.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/ld" -demangle -lto_library /Library/Developer/CommandLineTools/usr/lib/libLTO.dylib -dynamic -arch x86_64 -macosx_version_min 10.15.0 -syslibroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -o $WORK/b001/exe/a.out -L/usr/local/Cellar/librdkafka/1.3.0/lib -L/Users/s.rabot/git/go-lectra-maintenance/build/lib -headerpad 1144 -no_pie -pagezero_size 4000000 /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/go.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000000.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000001.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000002.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000003.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000004.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000005.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000006.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000007.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000008.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000009.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000010.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000011.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000012.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000013.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000014.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000015.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000016.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000017.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000018.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000019.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000020.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000021.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000022.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000023.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000024.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000025.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000026.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000027.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000028.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000029.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000030.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000031.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000032.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000033.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000034.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-745841879/000035.o -lrdkafka -lpthread -framework CoreFoundation -framework Security -lrdkafka -L/usr/local/lib -lSystem /Library/Developer/CommandLineTools/usr/lib/clang/11.0.0/lib/darwin/libclang_rt.osx.a
ld: warning: directory not found for option '-L/usr/local/Cellar/librdkafka/1.3.0/lib'
```

I also tried with `PKG_CONFIG_PATH` instead if `PKGCONFIG_PATH`. No luck.

```
PKG_CONFIG_PATH=/Users/s.rabot/git/go-lectra-maintenance/build/lib/pkgconfig GOOS=darwin GOARCH=amd64 go build -ldflags '-extldflags "-v -L/Users/s.rabot/git/go-lectra-maintenance/build/lib -lrdkafka"' -o dist/maintenance
# gitlab.lectra.com/cloud/maintenance
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin19.2.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/ld" -demangle -lto_library /Library/Developer/CommandLineTools/usr/lib/libLTO.dylib -dynamic -arch x86_64 -macosx_version_min 10.15.0 -syslibroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -o $WORK/b001/exe/a.out -L/usr/local/Cellar/librdkafka/1.3.0/lib -L/Users/s.rabot/git/go-lectra-maintenance/build/lib -headerpad 1144 -no_pie -pagezero_size 4000000 /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/go.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000000.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000001.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000002.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000003.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000004.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000005.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000006.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000007.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000008.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000009.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000010.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000011.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000012.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000013.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000014.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000015.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000016.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000017.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000018.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000019.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000020.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000021.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000022.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000023.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000024.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000025.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000026.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000027.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000028.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000029.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000030.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000031.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000032.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000033.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000034.o /var/folders/z0/rcywz4nn3jg6g96h67c_4xzdw31lvl/T/go-link-583248584/000035.o -lrdkafka -lpthread -framework CoreFoundation -framework Security -lrdkafka -L/usr/local/lib -lSystem /Library/Developer/CommandLineTools/usr/lib/clang/11.0.0/lib/darwin/libclang_rt.osx.a
ld: warning: directory not found for option '-L/usr/local/Cellar/librdkafka/1.3.0/lib'
```

Here  `pkg-config --libs rdkafka` output with `PKG_CONFIG_PATH`:

```
$ PKG_CONFIG_PATH=/Users/s.rabot/git/go-lectra-maintenance/build/lib/pkgconfig pkg-config --libs rdkafka
-L/Users/s.rabot/git/go-lectra-maintenance/build/lib -lrdkafka
```

**Comment by edenhill** [comment]
What happens if you uninstall the brew librdkafka?

**Comment by edenhill** [comment]
Also, do you get an actual error in the build? That ld warning can be ignored.

**Comment by sylr** [comment]
librdkafka does not exists anymore in brew, I uninstalled it when I decided to ship librdkafka with the project. I checked, I can't find any reference of it in '/usr/local' any more.

You're right, I was so dived in that at one point I made it compile but just assumed it did not succeed because of the warning.

FYI just `GOOS=darwin GOARCH=amd64 go build -ldflags '-extldflags "-v -L/Users/s.rabot/git/go-lectra-maintenance/build/lib -lrdkafka"' -o dist/maintenance` works, no need for PKG_CONFIG_PATH.

I'm still confused about the '-L/usr/local/Cellar/librdkafka/1.3.0/lib' warning.
